pkg/crc/api/api.go

```diff
@@ -69,12 +69,34 @@ func (api CrcApiServer) handleConnection(conn net.Conn) {
 		logging.Error("Error decoding request: ", err.Error())
 		return
 	}
-
-	if handler, ok := api.handlers[req.Command]; ok {
-		result := handler(req.Args)
-		writeStringToSocket(conn, result)
+	if req.Command == "start" || req.Command == "stop" || req.Command == "delete" {
+		// queue new request to channel
+		if !addRequestToChannel(req, requests) {
+			defer conn.Close()
+			logging.Error("Channel capacity reaced, unable to add new request")
+			writeStringToSocket(conn, "Channel capacity reaced, unable to add new request")
+			return
+		}
+		go func(requests <-chan commandRequest) {
+			req := <-requests
```

Comment:
Since we spawn one goroutine each time we add an item to the channel, I don't think this will help serialize start/stop/delete requests, they will still potentially be run concurrently. I'm only talking about this goroutine/channel use, which maybe could be improved. I know you've also added mutexes

Reply:
Yes, so the channel is used as a queue to add the incoming requests, this makes sure the requests are served in order, and the mutex lets only one of start/stop/delete handler to run at a time.

Reply:
I don't think this is working as intended.
Let's say you receive 2 requests, A and B.
First request gets added to the queue, a go routine is spawned
Second request gets added to the queue, a second go routine is spawned

One of the go routine will read from the channel and get request A
The other go routine will read from the channel and get request B
However, I don't think we have anything ensuring which go routine will call its handler first, they will be racing with each other.

Reply:
@anjannath Can you verify this? Could just a mutex be enough?

Reply:
@gbraad I did look into this and also had a chat with teuf, what he explains in the comment does happen and the requests gets served out of order, he suggested to use a single go routine for the three cluster operations (start,stop,delete) in which case we don't need the mutex. Just pushed an update!